### ISSUE TITLE
hterm: titlebar character appearing over current line when typing if TERM=xterm-256color, scolling pane not working if TERM=hterm

### ISSUE THREAD

**Issue by aktau** [neovim/neovim#19040]
### Neovim version (nvim -v)

v0.8.0-dev gf479dd0bb

### Vim (not Nvim) behaves the same?

no, but neither does nvim defaults (though please read along)

### Operating system/version

Debian testing

### Terminal name/version

hterm [1.92.1](https://chromium.googlesource.com/apps/libapps/+/HEAD/hterm/doc/ChangeLog.md#1_92_1_2022_03_04_minor-bug-fixes) (as used by the chrome Secure Shell extension), it is meant to be [compatible](https://chromium.googlesource.com/apps/libapps/+/HEAD/nassh/doc/FAQ.md#what-is-hterm) with xterm

### $TERM environment variable

xterm-256color

### Installation

git clone && ...

### How to reproduce the issue

Warning: It does not occur on base `nvim --clean`. I'm in the process of creating a minimal .vimrc that reproduces this (see the end of the issue). But I thought I'd file this in case anyone looks at the asciicast and has an insight. If that doesn't qualify for having an open issue, feel free to close this and sorry for the ruckus, I'll get back if I ever get more to the bottom of it.

 It does not occur when using nvim in the native ChromeOS terminal instead of hterm (`$TERM` is also xterm-256color there).

The issue is reminiscent of https://github.com/neovim/neovim/issues/14650, with the difference that:
 
  1. My issue is triggering by typing, not scrolling.
  2. The characters that appear are not random. Instead they appear to be the terminal's titlebar.

This issue started occurring somewhen in the last two months AFAIK. I can partially fix it by setting (e.g.) `TERM=hterm`, despite that not being recommended by `:help $TERM`. However, that breaks pane scrolling (this can also be seen in the asciinema cast):

[![asciicast](https://asciinema.org/a/918ilG8xPwVQe5yHRWETZRCRU.svg)](https://asciinema.org/a/918ilG8xPwVQe5yHRWETZRCRU)

Terminal issues confound me, so any hints for debugging this would be welcome.

NOTE: hterm terminal encoding is UTF-8

### Expected behavior

I expect to see no stray characters that overlap my typing.

### Actual behavior

When I type, lots of characters start appearing over the line I'm writing.

### Minimal .vimrc

I'm trying to minimize this more. But it looks like both `set hidden` and `set title` are important to reproduce this. I've not been able to reproduce this without the nvim-cmp plugin. I'll have to dig in to see what it actually does.

```vim
" Install vim-plug if it's missing. Need curl(1) though.
if empty(glob('~/.vim/autoload/plug.vim'))
  silent !curl -fLo ~/.vim/autoload/plug.vim --create-dirs
    \ https://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim
  autocmd VimEnter * PlugInstall | source $MYVIMRC
endif

call plug#begin('~/.vim/bundle')

Plug 'hrsh7th/cmp-buffer'
Plug 'hrsh7th/nvim-cmp'

call plug#end()

set hidden                        " Handle multiple buffers better.
set title                         " Set the terminal's title

lua <<EOF
  local cmp = require('cmp')
  cmp.setup({
    snippet = { },
    mapping = cmp.mapping.preset.insert({ }),
    sources = cmp.config.sources({
      { name = 'buffer' },
    })
  })
EOF
```

**Comment by justinmk** [comment]
> This issue started occurring somewhen in the last two months AFAIK.

bisect? :)

**Comment by aktau** [comment]
Not until I find how hrsh7th/nvim-cmp is drawing its window. The plugin requires a very recent version of nvim due to usage of the `nvim_create_user_command` API (so it requires f94f75dc0512def7fbfdfe100eea2dab3352d61f). That CL was already bad.

**Comment by aktau** [comment]
At this point, I'm starting to think it's some [change](https://chromium.googlesource.com/apps/libapps/+/HEAD/hterm/doc/ChangeLog.md) in hterm in the last three months that provoked this, because I've rolled back both nvim and hrsh7th/nvim-cmp to mid 2021 or earlier.

**Comment by zeertzjq** [comment]
I suspect this is similar to #18573, caused by ncurses.

**Comment by aktau** [comment]
> I suspect this is similar to https://github.com/neovim/neovim/issues/18573, caused by ncurses.

I can reproduce this, but much less frequently than the asciinema cast in that issue. First I thought it didn't occur for me, but then it did after pressing `o` (newline and enter insert mode). However, after doing that:

 1. The glitch stays in place, on the line that I just created.
 2. I can't provoke it anymore.

A bit odd. It may be related.

**Comment by aktau** [comment]
I reported this to the hterm people as well (https://bugs.chromium.org/p/chromium/issues/detail?id=1338461). I had the idea of comparing  `infocmp -1` for `xterm-256color` and `hterm-256color`.

There I observed that `xterm-256color` and `xterm-256color` are very similar: https://gist.github.com/aktau/d493e6a50a1856a8b5de3d552b9544b0. In prettier format (- means missing from hterm-256color, + means that hterm-256color has it while xterm-256color doesn't):

| Variable numeric     | Cap-name | TCap code | Description                             |
| ---                  | ---      | ---       | ---                                     |
| - can_change         | ccc      | cc        | terminal can re- define existing colors |
| - status_line_esc_ok | eslok    | es        | escape can be used on the status line   |
| - has_status_line    | hs       | hs        | has extra status line                   |
| - dis_status_line    | dsl      | ds        | disable status line                     |
| - from_status_line   | fsl      | fs        | return from status line                 |
| - to_status_line     | tsl      | ts        | move to status line, column #1          |
| - initialize_color   | initc    | Ic        | initialize color #1 to (#2,#3,#4)       |
| + set_background     | setb     | Sb        | Set background color #1 pair to #1      |
| + set_foreground     | setf     | Sf        | Set foreground color #1                 |

I've left out the keypad codes which hterm-256color lacks because they're not used. Basically, the only impactful difference is that `hterm-256color` does not have  `eslok`, `hs`, `dsl`, `fsl`, `tcl` codes to manipulate the status (title) bar. **This explains why I don't see the corruption with title characters when using `TERM=hterm-256color`**. It does not explain why pane scrolling is broken when using `TERM=hterm-256color`, as there is noticeable cap difference.

Then I remembered that Neovim fixes up terminal capabilities, and it does so specifically for xterm-alikes, but not for hterm because it doesn't know about hterm. So I made a patch to make hterm more like xterm:

```diff
git diff | cat -
diff --git a/src/nvim/tui/tui.c b/src/nvim/tui/tui.c
index 6c1c1ade0..ba2cf29ac 100644
--- a/src/nvim/tui/tui.c
+++ b/src/nvim/tui/tui.c
@@ -1653,6 +1653,7 @@ static void patch_terminfo_bugs(TUIData *data, const char *term, const char *col
   bool xterm = terminfo_is_term_family(term, "xterm")
                // Treat Terminal.app as generic xterm-like, for now.
                || nsterm;
+  bool hterm = terminfo_is_term_family(term, "hterm");
   bool kitty = terminfo_is_term_family(term, "xterm-kitty");
   bool linuxvt = terminfo_is_term_family(term, "linux");
   bool bsdvt = terminfo_is_bsd_console(term);
@@ -1716,7 +1717,7 @@ static void patch_terminfo_bugs(TUIData *data, const char *term, const char *col
     unibi_set_bool(ut, unibi_back_color_erase, false);
   }
 
-  if (xterm) {
+  if (xterm || hterm) {
     // Termit, LXTerminal, GTKTerm2, GNOME Terminal, MATE Terminal, roxterm,
     // and EvilVTE falsely claim to be xterm and do not support important xterm
     // control sequences that we use.  In an ideal world, these would have
```

**NOTE**: This is not a complete patch, we should add `hterm` to another function too.

The patch basically makes the following lines get executed: https://github.com/neovim/neovim/blob/b20b28249b1c58698be09c5b9d6785cfb1b12fb3/src/nvim/tui/tui.c#L1740-L1752.
**This fixes the pane scrolling on `TERM=hterm-256color`**. I think this means that hterm doesn't implement pane scrolling, but the (new) hterm-256color terminfo wrongly includes it. This is an upstream bug that requires a Neovim fixup in the mean time.

But one thing still bothered me: why did the title flickering not come back when I made hterm have the same fixups as xterm (i.e.: with an expanded patch that adds `|| hterm` everywhere there is `xterm`)? They should be basically identical then, I reckoned. This is especially the case because Neovim actually **adds* the status capability to any xterm-alike (and also hterm-alike after my patch): https://github.com/neovim/neovim/blob/b20b28249b1c58698be09c5b9d6785cfb1b12fb3/src/nvim/tui/tui.c#L1728-L1729

Then why doesn't hterm show the same bad title flickering as when using xterm-256color? I couldn't find the answer, until I resorted in desparation to diffing the end result: the `unibi_term` state inside of Neovim:

  1. Start a Neovim with `TERM=xterm-256color` and `unibi_dump()` the `unibi_term`.
  2. Start a Neovim with `TERM=hterm-256color` and `unibi_dump()` the `unibi_term`.
  3. Load both of them using `unibi_from_mem()`.
  4. Print them out in a human-readable manner using macro-magic [1] (I know there's extension codes I'm missing, but I believe it's not important here). It would've been nice to have a `unibi_dump_human()` function or some such, but in the end not too painful to make.

The result of diffing that is: https://gist.github.com/aktau/834b09253c55dc21f64e2a8aaef461fb. Indeed, Neovim has patched the hterm-256color to contain status line codes, but they are different from the ones that  come with the (new) version of xterm-256color:

```diff
-unibi_dis_status_line: ~
+unibi_dis_status_line: (null)
-unibi_from_status_line: }
+unibi_from_status_line: ^G // patched by Neovim
-unibi_to_status_line: ~}%p1%d` // comes  from xterm-256color
+unibi_to_status_line: [m // patched by Neovim
```

So my current theory about why I don't see corruption when using hterm-256color **even** when I tell Neovim to apply the same adjustments to as it used to do when using xterm-256color, is that  hterm responds differently to the term codes. For example, hterm may ignore `[m` (normally patched by Neovim), but corrupts the screen with `~}%p1%d`.

Regardless of all that, it is wrong to fixup `hterm-256color` to contain any status line capabilities, and wasteful to boot (the amount of corruption on my screen indicates the title is updated far more often than it should be). So I'll add that to a future patch too.

Current status:

|                | without patch                                                    | with patch                                               |
| ----           | -----                                                            | ----                                                     |
| **xterm-256color** | title appears over characters, pane scrolling OK                 | /                                                        |
| **hterm-256color** | title does NOT appear over characters, pane scrolling CORRUPTION | title does NOT appear over characters, pane scrolling OK |

[1]:

```c
static void un_dump_human(unibi_term *u) {
  ILOG("name = %s", unibi_get_name(u));

#define print_bool(en) \
  ILOG(#en ": %d", unibi_get_bool(u, en));
  print_bool(unibi_auto_left_margin);
  print_bool(unibi_auto_right_margin);
  print_bool(unibi_no_esc_ctlc);
  print_bool(unibi_ceol_standout_glitch);
  print_bool(unibi_eat_newline_glitch);
  // ...
}
```
